src/opencl_bitcracker_fmt_plug.c

```diff
@@ -0,0 +1,630 @@
+    HANDLE_CLERROR(ciErr1, "clSetKernelArg"); 
+    ciErr1 = clSetKernelArg(prepare_kernel, 1, sizeof(cl_mem), (void*)&salt_d);
+    HANDLE_CLERROR(ciErr1, "clSetKernelArg"); 
+    ciErr1 = clSetKernelArg(prepare_kernel, 2, sizeof(cl_mem), (void*)&padding_d);
+    HANDLE_CLERROR(ciErr1, "clSetKernelArg"); 
+    ciErr1 = clSetKernelArg(prepare_kernel, 3, sizeof(cl_mem), (void*)&w_blocks_d);
+    HANDLE_CLERROR(ciErr1, "clSetKernelArg"); 
+    
+    szLocalWorkSize = get_kernel_max_lws(gpu_id, prepare_kernel);
+    szGlobalWorkSize = szLocalWorkSize*max_compute_units;
+	
+	ciErr1 = clEnqueueNDRangeKernel(
+									queue[gpu_id], prepare_kernel, 1, NULL, &szGlobalWorkSize, 
+									&szLocalWorkSize, 0, NULL, NULL
+								);
+    HANDLE_CLERROR(ciErr1, "clEnqueueNDRangeKernel"); 
+
+    ciErr1 = clEnqueueReadBuffer(	queue[gpu_id], w_blocks_d, CL_TRUE, 0, 
+    								BITCRACKER_SINGLE_BLOCK_SHA_SIZE*BITCRACKER_ITERATION_NUMBER*sizeof(unsigned int), 
+    								w_blocks_h, 0, NULL, NULL
+    							);
+    HANDLE_CLERROR(ciErr1, "clEnqueueNDRangeKernel"); 
+
+    HANDLE_CLERROR(clFlush(queue[gpu_id]), "clFlush"); 
+    HANDLE_CLERROR(clFinish(queue[gpu_id]), "clFinish"); 
+
+out:
+
+	HANDLE_CLERROR(clReleaseKernel(prepare_kernel), "[BitCracker] -> Release kernel w blocks");
+
+	clReleaseMemObject(salt_d);
+	clReleaseMemObject(padding_d);
+	clReleaseMemObject(w_blocks_d);     
+	free(padding);
+
+	return;
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{	
+	if(!ciphertext)
+	{
+		printf("\n[BitCracker] -> Error: target encrypted unit not found!\n");
+		return 0;
+	}
+
+	printf("\n[BitCracker] -> **** Extracting metadata from disk image \"%s\" **** \n", ciphertext);
+	diskImage = fopen(ciphertext, "r");
+	if (diskImage == NULL){
+		printf("[BitCracker] -> Failed to open memory area target, please check if device exists or if you have right permissions\n");
+		exit(1);
+	}
+	readData(diskImage);
+	printf("[BitCracker] -> ********************************************* \n\n");
+
+	return 1;
+}
+
+
+static char *split(char *ciphertext, int index, struct fmt_main *self)
+{
+	return ciphertext;
+}
+
+static void *get_binary(char *ciphertext)
+{
+	return (void *) ciphertext;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	return tmpIV;
+}
+
+static void set_salt(void *salt)
+{
+	memset(tmpIV, 0, BITCRACKER_IV_SIZE);
+	memcpy(tmpIV + 1, nonce, BITCRACKER_NONCE_SIZE);
+	if(BITCRACKER_IV_SIZE-1 - BITCRACKER_NONCE_SIZE - 1 < 0)
+	{
+		fprintf(stderr, "Nonce error\n");
+		return;
+	}
+	*tmpIV = (unsigned char)(BITCRACKER_IV_SIZE - 1 - BITCRACKER_NONCE_SIZE - 1);
+	tmpIV[BITCRACKER_IV_SIZE-1] = 1; 
+}
+
+void cpu_print_hex(unsigned char hash[], int size)
+{
+   int idx;
+   for (idx=0; idx < size; idx++)
+      printf("0x%02x,",hash[idx]);
+   printf("\n");
+}
+
+void fillBuffer(FILE *fp, unsigned char *buffer, int size){
+	int k;
+	for(k=0;k<size;k++){
+		buffer[k] = (unsigned char)fgetc(fp);
+	}
+}
+
+void readData(FILE * diskImage){
```

Comment:
This functionality should be moved to a separate `bitlocker2john.py` or `bitlocker2john.c` program.